Undefined behavior: "shift exponent X is too large"

Undefined behavior detected when Mlucas is compiled with UBSan enabled.

GCC:
```
../src/Mlucas.c:3083:58: runtime error: shift exponent 64 is too large for 64-bit type 'long long unsigned int'
../src/dft_macro.c:5057:39: runtime error: shift exponent 32 is too large for 32-bit type 'unsigned int'
../src/dft_macro.c:3279:38: runtime error: shift exponent 32 is too large for 32-bit type 'unsigned int'
```
Clang:
```
../src/Mlucas.c:3083:58: runtime error: shift exponent 64 is too large for 64-bit type 'uint64' (aka 'unsigned long long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/Mlucas.c:3083:58 in 
 ../src/dft_macro.c:5057:39: runtime error: shift exponent 32 is too large for 32-bit type 'uint32' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/dft_macro.c:5057:39 in 
../src/dft_macro.c:3279:38: runtime error: shift exponent 32 is too large for 32-bit type 'uint32' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/dft_macro.c:3279:38 in 
```
---
Clang Tidy also outputs similar warnings, although some of these may be false positives:
```
$ grep -A 2 clang-analyzer-core.BitwiseShift clang-tidy.log
 ../src/Mlucas.c:5558:28: warning: Left shift by '64' overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
 5558 |                                 curr_word = (curr_wd64 << (64 - bits[ii]));                     /* Off-shift everything but the bits we need... */
      |                                              ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
--
 ../src/Mlucas.c:5633:28: warning: Left shift by '64' overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
 5633 |                                 curr_word = (curr_wd64 << (64 - bits[ii]));                     /* Off-shift everything but the bits we need... */
      |                                              ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
--
 ../src/mi64.c:692:16: warning: Left shift overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
  692 |                 y[i] = (x[i] << nshift) + (x[i-1] >> m64bits);
      |                         ~~~~~^~~~~~~~~
--
 ../src/mi64.c:695:15: warning: Left shift overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
  695 |         y[0] = (x[0] << nshift);
      |                 ~~~~~^~~~~~~~~
--
 ../src/mi64.c:1146:16: warning: Right shift overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
 1146 |                 y[i] = (x[i] >> nshift) + (x[i+1] << m64bits);
      |                         ~~~~~^~~~~~~~~
--
 ../src/mi64.c:1149:23: warning: Right shift overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
 1149 |         y[len-1] = (x[len-1] >> nshift);
      |                     ~~~~~~~~~^~~~~~~~~
--
 ../src/mi64.c:3910:14: warning: Left shift overflows the capacity of 'unsigned long long' [clang-analyzer-core.BitwiseShift]
 3910 |                 j = ((1ull << exp)+63)>>6;
      |                       ~~~~~^~~~~~
--
 ../src/mi64.c:8374:20: warning: Right shift by '64' overflows the capacity of 'uint64' [clang-analyzer-core.BitwiseShift]
 8374 |                 lead_chunk = lo64>>(64-log2_numbits);
      |                              ~~~~^~~~~~~~~~~~~~~~~~~
--
 ../src/qfloat.c:546:4: warning: Right operand is negative in right shift [clang-analyzer-core.BitwiseShift]
  546 |                         QRSHIFT(qt, shft, qt);
      |                         ^
--
 ../src/twopmodq80.c:2374:14: warning: Right operand is negative in right shift [clang-analyzer-core.BitwiseShift]
 2374 |                 if((pshift >> j) & (uint32)1)
      |                     ~~~~~~~^~~~
```

--- a/src/Mlucas.c
+++ b/src/Mlucas.c
@@ -3100,7 +3100,7 @@ uint64 	shift_word(double a[], int n, const uint64 p, const uint64 shift, const 
 		// not including the current word, hence (64-mod64) rather than (63-mod64) as the BIGWORD_BITMAP[w64] shift count.
 		// NOTE: That (64 - ...) shiftcount also necessitates an added AND-mask to zero the result in the mod64 == 0 case:
 		itmp64 = BIGWORD_BITMAP[w64];	i = (int)bits_per_word;
-		ii = BIGWORD_NBITS[w64] + i*mod64 + popcount64( (itmp64<<(64-mod64)) & -(mod64 != 0) );
+		ii = BIGWORD_NBITS[w64] + i*mod64 + (mod64 ? popcount64( itmp64<<(64-mod64) ) : 0);
 		// Loop up or down (should be by at most 1 word in either direction) if the resulting #bits is such that RES_SHIFT maps to a <> word:
 		curr_wd_bits = i + ( (int64)(itmp64<<(63-mod64)) < 0 );
 		// Can gain a few % speed by commenting out this correction-step code, but even though I've encountered

--- a/src/Mlucas.c
+++ b/src/Mlucas.c
@@ -3098,7 +3098,6 @@ uint64 	shift_word(double a[], int n, const uint64 p, const uint64 shift, const 
 		// Then exactly compute the bitcount at the resulting word, by adding the BIGWORD_NBITS-array-stored exact
 		// total bitcount at the next-lower index-multiple-of-64 to the number of bits in the next (mod64) words,
 		// not including the current word, hence (64-mod64) rather than (63-mod64) as the BIGWORD_BITMAP[w64] shift count.
-		// NOTE: That (64 - ...) shiftcount also necessitates an added AND-mask to zero the result in the mod64 == 0 case:
 		itmp64 = BIGWORD_BITMAP[w64];	i = (int)bits_per_word;
 		ii = BIGWORD_NBITS[w64] + i*mod64 + (mod64 ? popcount64( itmp64<<(64-mod64) ) : 0);
 		// Loop up or down (should be by at most 1 word in either direction) if the resulting #bits is such that RES_SHIFT maps to a <> word:

--- a/src/dft_macro.c
+++ b/src/dft_macro.c
@@ -3274,7 +3274,7 @@ void RADIX_256_DIT(
 			tptr->re,tptr->im,(tptr+0x1)->re,(tptr+0x1)->im,(tptr+0x2)->re,(tptr+0x2)->im,(tptr+0x3)->re,(tptr+0x3)->im,(tptr+0x4)->re,(tptr+0x4)->im,(tptr+0x5)->re,(tptr+0x5)->im,(tptr+0x6)->re,(tptr+0x6)->im,(tptr+0x7)->re,(tptr+0x7)->im,(tptr+0x8)->re,(tptr+0x8)->im,(tptr+0x9)->re,(tptr+0x9)->im,(tptr+0xa)->re,(tptr+0xa)->im,(tptr+0xb)->re,(tptr+0xb)->im,(tptr+0xc)->re,(tptr+0xc)->im,(tptr+0xd)->re,(tptr+0xd)->im,(tptr+0xe)->re,(tptr+0xe)->im,(tptr+0xf)->re,(tptr+0xf)->im,
 			c16,s16	// These = (c16,s16) typically def'd for use in the radix-16 DFT
 		);	tptr += 16;
-		if(i_idx) {	// vvv +2 here because this is setup for next loop pass
+		if(i_idx && i < 15) {	// vvv +2 here because this is setup for next loop pass
 			nshift = i+i+2;	// i_idx shift counts here run as >>2,4,...,30
 			off_ptr = i_offsets_lo + ( ((i_idx>>nshift)&0x3) << 4 );
 			p0  = off_ptr[0x0];p1  = off_ptr[0x1];p2  = off_ptr[0x2];p3  = off_ptr[0x3];p4  = off_ptr[0x4];p5  = off_ptr[0x5];p6  = off_ptr[0x6];p7  = off_ptr[0x7];p8  = off_ptr[0x8];p9  = off_ptr[0x9];pa  = off_ptr[0xa];pb  = off_ptr[0xb];pc  = off_ptr[0xc];pd  = off_ptr[0xd];pe  = off_ptr[0xe];pf  = off_ptr[0xf];
@@ -4272,7 +4272,7 @@ in the same order here as DIF, but the in-and-output-index offsets are BRed: j1 
 		// Intermediates pointers:
 		vec_dbl *tmp, *v0,*v1,*v2,*v3,*v4,*v5,*v6,*v7,
 			// Each of these rhi-ptrs points to next 8 vec_cmplx= 16 vec_dbl data sets:
-			*r10 = r00+0x10,*r20 = r00+0x20,*r30 = r00+0x30,*r40 = r00+0x40,*r50 = r00+0x50,*r60 = r00+0x60,*r70 = r00+0x70,
+			*r10,*r20,*r30,*r40,*r50,*r60,*r70,
 			*w[14], **twid_ptrs = &(w[0]);	// Array of 14 SIMD twiddles-pointers for the reduced-#args version of SSE2_RADIX8_DIF_TWIDDLE_OOP
 		/* Addresses into array sections */
 		double *add0, *add1, *add2, *add3, *add4, *add5, *add6, *add7;
@@ -4441,6 +4441,11 @@ in the same order here as DIF, but the in-and-output-index offsets are BRed: j1 
 		} else {
 			ASSERT(sc_arr != 0, "This function requires an initial Init-consts-mode call (in 1-thread mode only) before use!");
 		}	/* end of inits */
+
+		// Each of these rhi-ptrs points to next 8 vec_cmplx = 16 vec_dbl data sets; done here (rather than
+		// in the above declarations) since r00 == 0x0 on the (thr_id == -1) init-mode calls above, and
+		// forming pointer offsets from a null base pointer is undefined behavior:
+		r10 = r00+0x10; r20 = r00+0x20; r30 = r00+0x30; r40 = r00+0x40; r50 = r00+0x50; r60 = r00+0x60; r70 = r00+0x70;
 
 		/* If multithreaded, set the local-store pointers needed for the current thread; */
 	#ifdef MULTITHREAD
@@ -5054,7 +5059,7 @@ in the same order here as DIF, but the in-and-output-index offsets are BRed: j1 
 				isrt2,two,
 				tm0,OFF1,OFF2,OFF3,OFF4
 			); tm0 += 32;
-			if(i_idx) {	// vvv +2 here because this is setup for next loop pass
+			if(i_idx && i < 15) {	// vvv +2 here because this is setup for next loop pass
 				nshift = i+i+2;	// i_idx shift counts here run as >>2,4,...,30
 				off_ptr = i_offsets_lo + ( ((i_idx>>nshift)&0x3) << 4 );
 			}

